### PR TITLE
BUG: regression when applying groupby aggregation on categorical columns

### LINKED ISSUE #31256: Regression bugs when applying GroupBy Aggregations to Categorical columns
```python
>>> import pandas as pd

>>> ids = range(5)
>>> groups = pd.Series([0, 1, 1, 2, 2])
>>> values = pd.Categorical([0, 0, 0, 0, 1])
>>> df = pd.DataFrame({
            'id': ids,
            'groups': groups,
            'value': values
    }).set_index('id')

>>> df.groupby('groups').agg(pd.Series.nunique)
       value
groups
0          1
1          1
2        NaN

>>> df.groupby('groups').agg(pd.Series.nunique)['value'].dtype
CategoricalDtype(categories=[0, 1], ordered=False)

>>> df.groupby('groups').agg(pd.Series.count)
       value
groups
0          1
1        NaN
2        NaN

>>> df.groupby('groups').agg(pd.Series.mode)
                                                                           value
groups
0                                                                              0
1              0    0
Name: value, dtype: category
Categories (2, int64): [0, 1]
2       0    0
1    1
Name: value, dtype: category
Categories (2, int64): [0, 1]

```
#### Problem description
There's some regression bugs in `1.0.0rc0` when applying groupby aggregations to Categorical columns. All of these examples work as expected in `0.25.3`. 

Applying `pd.Series.nunique` and `pd.Series.count` both result in `NaN` for some groupings, the result has a Categorical dtype. The resulting dtype should be int64 and `NaN` should not be produced.

Applying `pd.Series.mode` returns the correct result for the first grouping but produces nonsensical output for subsequent groupings. 

#### Expected Output
Aggregations should work as expected on Categorical columns as they do in `0.25.3`:
```python
>>> df.groupby('groups').agg(pd.Series.nunique)
        value
groups
0           1
1           1
2           2

>>> df.groupby('groups').agg(pd.Series.nunique)['value'].dtype
dtype('int64')

>>> df.groupby('groups').agg(pd.Series.count)
        value
groups
0           1
1           2
2           2

>>> df.groupby('groups').agg(pd.Series.mode)
         value
groups
0      0     0
1      0     0
2      0     0
       1     1
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.9.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.0rc0
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.1
setuptools       : 45.1.0
Cython           : None
pytest           : 5.2.0
hypothesis       : None
sphinx           : 2.0.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.2.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : 0.3.2
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.0.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.2.0
s3fs             : 0.4.0
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : 0.47.0

</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [ ] closes #31256 
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by TomAugspurger** [comment]
@jbrockmendel did you have thoughts on this approach?

File: pandas/core/arrays/base.py
Comment (jbrockmendel) [inline pandas/core/arrays/base.py]: why is this change necessary?

File: pandas/core/groupby/groupby.py
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py]: up until a moment ago this had a length check; without that check this looks like it isnt actually changing anything

**Comment by jbrockmendel** [comment]
> did you have thoughts on this approach?

ATM just confusion.  Pending the CI, I'll take @charlesdong1991 's word for it that just changing `dtype=dtype` to `dtype=dtype.name` fixes the new test, but it isn't at all clear to me why that would work.

File: pandas/core/groupby/groupby.py
Comment (charlesdong1991) [reply pandas/core/groupby/groupby.py]: in the previous commit, i did add a length check, but i do not feel it was right solution. the issue is `dtype` preserves its categories @jbrockmendel

Comment (charlesdong1991) [reply pandas/core/groupby/groupby.py]: so `dtype.categories` preserves the original value, in this case, [0, 1] as its categories, however, after aggregation, we should NOT use the original categories, but the new one based on the result

File: pandas/tests/groupby/test_categorical.py
Comment (charlesdong1991) [inline pandas/tests/groupby/test_categorical.py]: I think the output of this one was wrong, the aggregated result should not preserve its original categories?

currently on master, this will return `CategoricalDtype(categories=['first', 'fourth', 'second', 'third'], ordered=True)`, but i think this should be `CategoricalDtype(categories=['second', 'third'], ordered=True)`

@jbrockmendel

**Comment by TomAugspurger** [comment]
I guess I'm not sure what to do here.

**Comment by TomAugspurger** [comment]
I'm not really sure what to do here. There's no guarantee that `func(x).dtype == x.dtype` for user-defined functions.

We're relying on the heuristic that if `ExtensionArray._from_sequence(func(x))` works without exception, then `func(x).dtype` should be `x.dtype`. But that's not a valid assumption...

**Comment by TomAugspurger** [comment]
One question (to anyone, including myself): why are we calling `_try_cast` on the result of user-defined functions?  The comment about "perhaps going through object dtype" don't seem to apply to this. If the user-defined function goes through object dtype then that's on them, right?

**Comment by TomAugspurger** [comment]
Removing that try cast shows a handful of failures

```
pandas/tests/groupby/test_function.py::test_median_empty_bins[True] FAILED                                                                                              [  5%]
pandas/tests/groupby/test_function.py::test_ops_general[count-size] FAILED                                                                                              [ 10%]
pandas/tests/groupby/test_groupby.py::test_with_na_groups[float32] FAILED                                                                                               [ 15%]
pandas/tests/groupby/test_groupby.py::test_with_na_groups[int64] FAILED                                                                                                 [ 21%]
pandas/tests/groupby/test_groupby.py::test_with_na_groups[int32] FAILED                                                                                                 [ 26%]
pandas/tests/groupby/test_groupby.py::test_with_na_groups[int16] FAILED                                                                                                 [ 31%]
pandas/tests/groupby/test_groupby.py::test_with_na_groups[int8] FAILED                                                                                                  [ 36%]
pandas/tests/groupby/aggregate/test_aggregate.py::test_aggregate_item_by_item FAILED                                                                                    [ 42%]
pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-mean-mean] FAILED                                                                     [ 47%]
pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-median-<lambda>] FAILED                                                               [ 52%]
pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-var-<lambda>] FAILED                                                                  [ 57%]
pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[NotImplementedError] FAILED                                                                 [ 63%]
pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[RuntimeError] FAILED                                                                        [ 68%]
pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[KeyError] FAILED                                                                            [ 73%]
pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[IndexError] FAILED                                                                          [ 78%]
pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[OSError] FAILED                                                                             [ 84%]
pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ValueError] FAILED                                                                          [ 89%]
pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ArithmeticError] FAILED                                                                     [ 94%]
pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[AttributeError] FAILED                                                                      [100%]
```

The first few look like buggy tests I think. For example, why is this a float result (behavior on 0.25.3 / master)?

```
In [15]: pd.DataFrame({"A": [1., 2.]}).groupby([1, 1]).agg(np.size)
Out[15]:
     A
1  2.0
```

Regardless, I'm straying from what can reasonable be changed for the release. What should we do short-term? Do we want to restore 0.25.3 behavior? I'm not even sure how feasible that is. It seems to be taking a very different code path compared to 0.25.3 (based on a quick skim).

**Comment by jbrockmendel** [comment]
> why are we calling _try_cast on the result of user-defined functions?

IIRC from refactoring a few months ago, some of the usages were about converting floats back to ints, which we do elsewhere and is fine.  The rest I think is kludges for datetimelike ops that cast to i8 before calling cython, and then _should_ case back immediately but dont.

**Comment by jbrockmendel** [comment]
> We're relying on the heuristic that if ExtensionArray._from_sequence(func(x)) works without exception

I've been thinking it would be useful to have either a kwarg for _from_sequence or another method that is only goes through if that would be the inferred dtype, e.g. a datetime64[ns] ndarray would work for DatetimeArray._from_sequence(arr, strict=True), but a int64 ndarray wouldnt.

**Comment by jbrockmendel** [comment]
> Removing that try cast shows a handful of failures

Removing only the EA part of _try_cast trims down those failures to all be on test_groupby_agg_err_catching, and all returning object dtype when we expect decimal dtype.

**Comment by charlesdong1991** [comment]
> The expected dtype of `SeriesGroupBy[categorical].count()` should be int64, not categorical. IIUC, this change returning a Categorical whose categories are the _counts_, which I don't think we want. It seems to me like we shouldn't be trying to cast back to an EA here. We know that count shouldn't cast back, but for arbitrary user-defined functions we don't know that.

I definitely agree on it, this should be int64, not original dtype of `value` column, and in this case, `Categorical`. At the first place, I thought should just not to cast it to EA, but i did that, and lots of tests failed, very similar to what you showed. And for some cases like count or size, the result should not be casted, and I did not find a clean way to avoid `_try_cast`.

Then reading through those failed cases, seems the dtype of agg result are aligned with dtype of that column (by casting)

> The first few look like buggy tests I think. For example, why is this a float result (behavior on 0.25.3 / master)?
> 
> ```
> In [15]: pd.DataFrame({"A": [1., 2.]}).groupby([1, 1]).agg(np.size)
> Out[15]:
>      A
> 1  2.0
> ```

and this explains why the result is a float, because the agg column A is float type, and this is casted as well by `_try_cast`. I also have doubt if this is desired behaviour or not, since ideally, like `count`, this should return an `int`. But seems to me, there is quite difficult to restore 0.25.3 behaviour, and a lot new features might be lost if doing so.  

Also in `test_decimal.py`, i think there were a couple user defined functions, and the result of agg is inferred to decimal dtype .

**Comment by TomAugspurger** [comment]
> some of the usages were about converting floats back to ints, which we do elsewhere and is fine

Do you recall the details here? I don't think that `pd.DataFrame({"A": [1., 2.]}).groupby([1, 1]).agg(np.size)` returning floats is correct (which we do on 0.25.3 and master).

> have either a kwarg for _from_sequence

I think that's the original intent of `_from_sequence`, to only accept a sequence of instances of the scalar.

---

I think solving this properly will need some kind of indication from the user... There are fundamentally ambiguous cases.

**Comment by TomAugspurger** [comment]
I may have a potential halfway fix, that may be appropriate for 1.0.0. Will push it here in a bit.

**Comment by TomAugspurger** [comment]
I pushed that change here. The basic rule is we cast back to an EA when the scalars are instances of the ExtensionDtype's type.

This required changing a few tests. I think the updated tests are a bit more correct. Can people (cc @jreback @jorisvandenbossche @jbrockmendel @WillAyd) take a look and let me know if we want to proceed with this route for 1.0.0?

**Comment by WillAyd** [comment]
I’m not a huge fan of the current fix - do we know where the regression was introduced?

File: pandas/core/groupby/groupby.py
Comment (jreback) [inline pandas/core/groupby/groupby.py]: can u do this inside try_to_cast_ea?

i am ok with this generally

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
I recently opened an issue that is related to this topic of "trying to cast back" and some problem with that: https://github.com/pandas-dev/pandas/issues/31108

> >  have either a kwarg for _from_sequence
>
> I think that's the original intent of _from_sequence, to only accept a sequence of instances of the scalar.

Yes, that was indeed the intent. Our own implementation of `_from_sequence` are IMO buggy because we are using them for a lot of other things as well. They should only accept instances of its own scalar type.

> I think solving this properly will need some kind of indication from the user... There are fundamentally ambiguous cases.

Yes, that sounds as a good idea. I think long term, we should add a way to specify the expected dtype of the result (and also for the plain DataFrame apply). I think in dask something like that is already done?

File: pandas/tests/groupby/test_categorical.py
Comment (jorisvandenbossche) [inline pandas/tests/groupby/test_categorical.py]: Can you also add a version with a non-numeric categories? (as that fails even more on master (all NaN), since none of the counts are possible categories)

**Comment by jorisvandenbossche** [comment]
> The basic rule is we cast back to an EA when the scalars are instances of the ExtensionDtype's type.

A potential problem with this is that it's not very robust / deterministic if you're only checking the first element (eg the first element could be a missing value, which is almost never an instance of the dtype's type)

**Comment by TomAugspurger** [comment]
> I’m not a huge fan of the current fix - do we know where the regression was introduced?

I'm not sure, but I think it's some combination of https://github.com/pandas-dev/pandas/pull/24488/files#diff-6d2a4be351f3d7ec35b9c6696a615242L771, and calling `_try_cast_result` in more places.

>  A potential problem with this is that it's not very robust / deterministic if you're only checking the first element (eg the first element could be a missing value, which is almost never an instance of the dtype's type)

Yep, good point. On balance though, I this is the best we can do for 1.0. It's slightly more complicated, but perhaps we check the first non-NA result?

I'll add an xfailing case.

**Comment by TomAugspurger** [comment]
There's a couple more test failures around resampling.

```
    def test_resample_integerarray():
        # GH 25580, resample on IntegerArray
        ts = pd.Series(
            range(9), index=pd.date_range("1/1/2000", periods=9, freq="T"), dtype="Int64"
        )
        result = ts.resample("3T").sum()
        expected = Series(
            [3, 12, 21],
            index=pd.date_range("1/1/2000", periods=3, freq="3T"),
            dtype="Int64",
        )
>       tm.assert_series_equal(result, expected)
E       AssertionError: Attributes of Series are different
E       
E       Attribute "dtype" are different
E       [left]:  float64
E       [right]: Int64
```

I think that `IntegerArray.resample().sum()` should be Int64 dtype, since IntegerArray._reduce('sum') is known to be an int. Fixing that is ugly, but doable.

The test also asserts that `IntegerArray.resample().mean()` is also Int64Dtype. I think that's incorrect, as `IntegerArray._reduce("mean")` is a float.

File: pandas/core/groupby/groupby.py
Comment (TomAugspurger) [reply pandas/core/groupby/groupby.py]: In principle I agree. For now (1.0.0) I think I prefer keeping this change narrow.

I'm definitely going to open a followup issue with some more discussion / potential fixes.

File: pandas/core/groupby/ops.py
Comment (TomAugspurger) [inline pandas/core/groupby/ops.py]: Ugly! But this ensures that `Series[IntDtype].resample().sum()` has int dtype. Assuming we want the restriction "only try casting back to EA when the scalars results are instances of the EA dtypes type, then I think this is the least worst option.

I briefly tried avoiding the cast from Int64 -> float with NaN, but that wasn't feasible today. Our Cython reductions like `group_add` only handle floats IIUC.

File: pandas/tests/resample/test_datetime_index.py
Comment (TomAugspurger) [inline pandas/tests/resample/test_datetime_index.py]: Changed the expected dtype. I think the old test was incorrect, as `mean` should be float.

File: pandas/tests/resample/test_timedelta.py
Comment (TomAugspurger) [inline pandas/tests/resample/test_timedelta.py]: Another test change. The expected result here is less clear I think. You could argue for either categorical or not...

**Comment by TomAugspurger** [comment]
>> A potential problem with this is that it's not very robust / deterministic if you're only checking the first element (eg the first element could be a missing value, which is almost never an instance of the dtype's type)

> Yep, good point. On balance though, I this is the best we can do for 1.0. It's slightly more complicated, but perhaps we check the first non-NA result?

I wasn't able to do this for Int64 dtype, even checking just non-NA values. The presence of an NA in `result` forced object dtype, so we had a mix of NA and Python ints. Python ints aren't instances of Int64Dtype.type (numpy.int64).

I'll push an xfailing test, but for now I'd recommend accepting this limitation in the hope that most reductions don't produce NA, and we're hoping to have a more robust fix later on.

File: pandas/tests/resample/test_timedelta.py
Comment (TomAugspurger) [reply pandas/tests/resample/test_timedelta.py]: This is an API change from 0.25.x. Dunno what to do here :/

Comment (TomAugspurger) [reply pandas/tests/resample/test_timedelta.py]: On 0.25.x, we have the following

```python
In [42]: df = pd.DataFrame({"A": ['a', 'b']}, dtype='category', index=pd.date_range('2000', periods=2))

In [43]: df.resample("2D").agg(lambda x: 'a')
Out[43]:
   A
0  a

In [44]: df.resample("2D").agg(lambda x: 'c')
Out[44]:
     A
0  NaN
```

We're potentially changing the *values* of the result by returning a CategoricalDtype there. I don't think that behavior is correct, so I'm in favor of making this breaking change... I think...

**Comment by TomAugspurger** [comment]
Pushed a release note. Can you take another look @jorisvandenbossche @jreback?

@WillAyd I agree that this fix isn't optimal. Do you think it's an improvement over the behavior on master though? Do you think it's OK for 1.0.0?

**Comment by WillAyd** [comment]
I’ll defer to others on this - don’t have a ton of time this week so don’t want to be a blocker.

Thanks for pushing the release

File: pandas/tests/resample/test_datetime_index.py
Comment (jreback) [reply pandas/tests/resample/test_datetime_index.py]: hmm, no it tries to cast back originally (which i suppose is dubious), but yeah we should *also* return float on nullable mean ops (thought we did genreally)

Comment (jorisvandenbossche) [reply pandas/tests/resample/test_datetime_index.py]: > I think the old test was incorrect, as mean should be float.

Yes, I agree that should be float.

File: pandas/tests/resample/test_timedelta.py
Comment (jorisvandenbossche) [reply pandas/tests/resample/test_timedelta.py]: > We're potentially changing the values of the result by returning a CategoricalDtype there. I don't think that behavior is correct, so I'm in favor of making this breaking change... I think...

That seems exactly the bug we were fixing initially. For groupby it was a regression, so fine to do it as a fix for resample I think.

**Comment by jorisvandenbossche** [comment]
@TomAugspurger for the whatsnew note you are now focussing on resample. Can you explain a bit how it comes that there is such a change for resample and not for groupby? I don't see much resample specific change in the diff?

File: doc/source/whatsnew/v1.0.0.rst
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.0.0.rst]:
```suggestion
The rules for the result dtype in :meth:`DataFrame.resample` aggregations have changed for extension types (:issue:`31359`).
```

**Comment by TomAugspurger** [comment]
Will double check, but the behavior change from 0.25.x is just for resample. The change from master affects both. Make sense?

File: doc/source/whatsnew/v1.0.0.rst
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.0.0.rst]: What will this return? Not a categorical? 
I am not fully sure that is correct as well. Eg if you do a first-like aggregation `lambda x: x[0]`, you would expect this to work... 

(but of course any heuristic will never be correct in all cases .. for full control in cases like this we will need an additional keyword)

**Comment by jorisvandenbossche** [comment]
> he behavior change from 0.25.x is just for resample. The change from master affects both. Make sense?

Yes, that's how I understood it. I just find it strange from looking at the diff ;)

File: pandas/tests/groupby/test_categorical.py
Comment (jorisvandenbossche) [inline pandas/tests/groupby/test_categorical.py]: I would also do the `agg(pd.Series.nunique)` here. The nunique method takes a different path (and was actually working fine on master)

**Comment by charlesdong1991** [comment]
Thanks for fixing and pushing the release! @TomAugspurger 👍 

Maybe I should not disrupt you, I just saw the changes you've made and maybe it's nicer to also add `*pandas 1.0.0*` above line 670 like you did on line 651 ^^

**Comment by TomAugspurger** [comment]
> nicer to also add *pandas 1.0.0* above line 670 like you did on line 651 ^^

Fixed, thanks!

File: pandas/tests/resample/test_timedelta.py
Comment (jorisvandenbossche) [reply pandas/tests/resample/test_timedelta.py]: I see now that I commented about in two different places (also in the whatsnew), with a contradicting message ...

I don't know what is best for now in this PR, but in general: if we are going to take the rule of "only trying to cast back if the scalars are of the dtype.type", then I think we should try to cast to categorical (meaning, we should see the string "a" as a valid scalar of a categorical dtype with string categories). The specific case of strings is a bit difficult of course, as also the categories' dtype is object and can hold anything.

Now, long term, I am thinking we should maybe not use this "scalar of dtype.type" rule (we can say that to explain it, but not as actual check in the code). But we could rather dictate that `_from_sequence` should be strict. And then it's `_from_sequence` that can hold the logic to know if it's a proper scalar (eg to know that pd.NA is also fine, or to know that "a" is a proper scalar for a categorical with categories that include "a", or knowing that a python int is also fine in addition to np.int64 scalar for an IntDtype, etc). 
For the specific case of categorical, it could then fail if it's getting scalars (eg "c" in the example above) which cannot be a scalar of its dtype (since the categories are part of the dtype, it can know "c" is not a valid scalar in that case). 

You will still get variable behaviour depending on what the agg function returns, eg `.agg(lambda x: 'a')` you get back categorical and `.agg(lambda x: 'c')` not, while both are simple functions that return a string. But with any rule such small differences will be unavoidable for the *default* behaviour (hence having a keyword to have full control might be interesting to investigate)

Comment (jorisvandenbossche) [reply pandas/tests/resample/test_timedelta.py]: Sorry for the long comment here :) That should probably be moved to the follow-up issue.  
About what is best for *this* PR, I am not fully sure. I suppose what you have now is fine. The `isinstance(result[0], dtype.type)` check just doesn't really work for categoricals ..

File: doc/source/whatsnew/v1.0.0.rst
Comment (TomAugspurger) [reply doc/source/whatsnew/v1.0.0.rst]: Right, it will be object dtype. That's the API breaking change. But, I'm OK with it because

1. It matches the behavior of groupby on 0.25.3. `df.groupby([1, 1]).agg(lambda x: 'a').A.dtype` is object.
2. It's incorrect (or at least surprising?) when the dtype can't hold the values (the next example where you get NaN values).

**Comment by TomAugspurger** [comment]
Good to merge when CI passes? I think this is the final thing before 1.0.

**Comment by jbrockmendel** [comment]
> Good to merge when CI passes?

Will take another look now

File: pandas/core/groupby/ops.py
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: TODO comment suggesting this get cleaned up, pointing back to this thread?

**Comment by jbrockmendel** [comment]
No objections to merging, agreed that some follow-up discussion is necessary on _from_sequence

File: pandas/core/groupby/ops.py
Comment (TomAugspurger) [reply pandas/core/groupby/ops.py]: Going to open an issue and will point here.

File: pandas/tests/resample/test_timedelta.py
Comment (TomAugspurger) [reply pandas/tests/resample/test_timedelta.py]: Sorry, missed this earlier. I'm writing up the followup issue now, but will address this there.

**Comment by TomAugspurger** [comment]
All green. Merging.

Thanks for kicking this off @charlesdong1991!

**Comment by TomAugspurger** [comment]
Handling the backport.

**Comment by charlesdong1991** [comment]
Thanks for your fix and help to release 1.0 for community! ALL credit to you!! 👍

File: pandas/core/groupby/groupby.py
Comment (simonjayhawkins) [inline pandas/core/groupby/groupby.py]: @charlesdong1991 it looks like this change caused the regression in #32194

**Comment by simonjayhawkins** [comment]
> > I’m not a huge fan of the current fix - do we know where the regression was introduced?
> 
> I'm not sure, but I think it's some combination of https://github.com/pandas-dev/pandas/pull/24488/files#diff-6d2a4be351f3d7ec35b9c6696a615242L771, and calling `_try_cast_result` in more places.

08087d6712b06502dc0f591b9f66987eba2bb220 is the first bad commit
commit 08087d6712b06502dc0f591b9f66987eba2bb220
Author: jbrockmendel <jbrockmendel@gmail.com>
Date:   Wed Nov 6 13:25:06 2019 -0800

    BUG: fix TypeErrors raised within _python_agg_general (#29425)

**Pull request merged**
